core/src/test/java/io/forestframework/core/http/HandlerIntegrationTest.java

```diff
@@ -0,0 +1,458 @@
+        throw new HttpException(HttpStatusCode.CLIENT_ERROR);
+    }
+}
+
+@Router("/preHandlerNotReturnBoolean")
+class ThrowExceptionsWhenPreHandlersNotReturnBoolean extends AbstractHandlerIntegrationTest {
+
+    @PreHandler("/**")
+    public String preHandler(HttpServerRequest request) {
+        runUnderTrace("prehandlerReturnsString");
+        return "shouldNotReturnString";
+    }
+
+    @Route(value = "/**", type = RoutingType.HANDLER)
+    public void handler(HttpServerRequest request, HttpServerResponse response) {
+        runUnderTrace("handler");
+    }
+
+    @OnError(value = "/**")
+    public void customErrorHandler(HttpServerResponse response, Throwable e) {
+        response.write(" should be handled by custom error handler. ");
+        response.write(" e.getMessage(): " + e.getMessage());
+        runUnderTrace("customErrorHandler");
+    }
+
+    @PostHandler("/**")
+    public void postHandler(HttpServerResponse response) {
+        runUnderTrace("postHandler");
+    }
+}
+
+class AbstractHandlerIntegrationTest {
+    List<String> traces = Collections.synchronizedList(new ArrayList<>());
+
+    void runUnderTrace(String traceId) {
+        traces.add(traceId);
+    }
+}
+
+@ExtendWith(ForestExtension.class)
+@ForestTest(appClass = MyCustomErrorHandlerApp.class)
+@DisableAutoScan
+@IncludeComponents(classes = {
+        MyCustom500HandlerRouter.class,
+        ErrorInCustomErrorHandler2.class,
+        Custom400ErrorHandler.class,
+        Uncaught404Error.class,
+        PrehandlerErrorPreventPropagation.class,
+        ErrorInPostHandler.class,
+        ThrowExceptionsWhenPreHandlersNotReturnBoolean.class,
+        PreHandlerReturnsVoidTrueOrFalse.class
+})
+public class HandlerIntegrationTest {
+    @Inject
+    MyCustom500HandlerRouter custom500HandlerRouter;
+
+    @Inject
+    Custom400ErrorHandler custom400ErrorHandler;
+
+    @Inject
+    ErrorInCustomErrorHandler2 errorInCustomErrorHandler;
+
+    @Inject
+    Uncaught404Error uncaught404Error;
+
+    @Inject
+    PrehandlerErrorPreventPropagation prehandlerErrorPreventPropagation;
+
+    @Inject
+    ErrorInPostHandler errorInPostHandler;
+
+    @Inject
+    ThrowExceptionsWhenPreHandlersNotReturnBoolean throwExceptionsWhenPreHandlersNotReturnBoolean;
+
+    @Inject
+    PreHandlerReturnsVoidTrueOrFalse preHandlerReturnsVoidTrueOrFalse;
+
+    @Inject
+    @Config("forest.http.port")
+    Integer port;
+
+    CloseableHttpClient client;
+
+    @BeforeEach
+    void cleanUpTraces() {
+        client = HttpClients.createDefault();
+        custom500HandlerRouter.traces.clear();
+        custom400ErrorHandler.traces.clear();
+        errorInCustomErrorHandler.traces.clear();
+        uncaught404Error.traces.clear();
+        prehandlerErrorPreventPropagation.traces.clear();
+        errorInPostHandler.traces.clear();
+        throwExceptionsWhenPreHandlersNotReturnBoolean.traces.clear();
+        preHandlerReturnsVoidTrueOrFalse.traces.clear();
```

Comment:
Please try to remove these copy-paste boilerplate code, very error-prone when you forget to add one.

Reply:
'kay.

Reply:
```
abstract class AbstractTraceableRouter {
   List<String> traceIds;
}

class Router1 extends AbstractTraceableRouter {}
class Router2 extends AbstractTraceableRouter {}

abstract class AbstractHandlerIntegrationTest {
   protected AbstractTraceableRouter router;
   @BeforeEach 
   void clear() {
      router.traceIds.clear(); 
   }
}

class Router1Test extends AbstractHandlerIntegrationTest {
    @Inject
    void setRouter1(Router1 router) {
        this.router = router;
    }
}
```